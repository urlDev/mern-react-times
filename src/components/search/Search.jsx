import React from "react";
import { useDispatch, useSelector } from "utils/react-redux-hooks";
import { useHistory } from "react-router-dom";

import {
  closeSearchModal,
  fetchSearch,
  openSearchModal,
  toggleSearchModal,
} from "redux/actions/chart";

import SearchSrc from "assets/searchIcon.svg";

import SearchModal from "components/search-modal/SearchModal";

import { NavSearchContainer, SearchIcon } from "./Search.styles";

const Search = ({ icon }) => {
  const { open } = useSelector((chart) => chart.chart);
  const { width } = useSelector((news) => news.news);
  const [input, setInput] = React.useState("");

  const dispatch = useDispatch();
  const history = useHistory();

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(fetchSearch(input));
    history.push(`/search/${input}`);
    setInput("");
    dispatch(closeSearchModal());
  };

  return (
    <NavSearchContainer open={open}>
      <form
        onSubmit={handleSubmit}
        onClick={width < 768 ? () => dispatch(openSearchModal()) : null}
      >
        <input
          type="text"
          placeholder="Search for a company or symbol"
          onChange={(e) => {
            setInput(e.target.value);
            dispatch(fetchSearch(e.target.value));
          }}
          open={open}
          value={input}
        />
        <SearchIcon
          icon={icon}
          open={open}
          onClick={() => dispatch(toggleSearchModal())}
        >
          <img src={SearchSrc} alt="search icon" />
        </SearchIcon>
      </form>
      <SearchModal />
    </NavSearchContainer>
  );
};

export default Search;
