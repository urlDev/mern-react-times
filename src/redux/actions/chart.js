import axios from "axios";

export const FETCH_FOREX_SUCCESS = "FETCH_FOREX_SUCCESS";
export const FETCH_CHART_ERROR = "FETCH_CHART_ERROR";
export const SET_LOADING = "SET_LOADING";
export const SET_LOADING_TRUE = "SET_LOADING_TRUE";
export const CHANGE_MARKET_TYPE = "CHANGE_MARKET_TYPE";
export const CHANGE_MARKET_NAME = "CHANGE_MARKET_NAME";
export const SET_MARKET_DETAIL = "SET_MARKET_DETAIL";
export const SET_CHART_DATA = "SET_CHART_DATA";
export const CLEAN_CHART_DATA = "CLEAN_CHART_DATA";
export const SET_CHART_TIME_FRAME = "SET_CHART_TIME_FRAME";
export const SET_RATING = "SET_RATING";
export const SET_HOME_CHART_DATA = "SET_HOME_CHART_DATA";
export const CLEAN_STATE = "CLEAN_STATE";
export const GET_SEARCH_RESULTS = "GET_SEARCH_RESULTS";
export const SET_SEARCH_MARKET_DETAIL = "SET_SEARCH_MARKET_DETAIL";
export const CLEAR_SEARCH_RESULTS = "CLEAR_SEARCH_RESULTS";
export const OPEN_SEARCH_MODAL = "OPEN_SEARCH_MODAL";
export const CLOSE_SEARCH_MODAL = "CLOSE_SEARCH_NODAL";

export const closeSearchModal = () => ({
  type: CLOSE_SEARCH_MODAL,
});

export const openSearchModal = () => ({
  type: OPEN_SEARCH_MODAL,
});

export const clearSearchResults = () => ({
  type: CLEAR_SEARCH_RESULTS,
});

export const setSearchMarketDetail = (details) => ({
  type: SET_SEARCH_MARKET_DETAIL,
  payload: details,
});

export const getSearchResults = (result) => ({
  type: GET_SEARCH_RESULTS,
  payload: result,
});

export const setLoading = () => ({
  type: SET_LOADING,
});

export const setLoadingTrue = () => ({
  type: SET_LOADING_TRUE,
});

export const cleanState = () => ({
  type: CLEAN_STATE,
});

export const setHomeChartData = (chart) => ({
  type: SET_HOME_CHART_DATA,
  payload: [chart],
});

export const setRating = (rating) => ({
  type: SET_RATING,
  payload: rating,
});

export const setChartTimeFrame = (time) => ({
  type: SET_CHART_TIME_FRAME,
  payload: time,
});

export const setChartData = (data) => ({
  type: SET_CHART_DATA,
  payload: data,
});

export const cleanChartData = () => ({
  type: CLEAN_CHART_DATA,
});

export const setMarketDetail = (market) => ({
  type: SET_MARKET_DETAIL,
  payload: market,
});

export const changeMarketType = (market) => ({
  type: CHANGE_MARKET_TYPE,
  payload: market,
});

export const changeMarketName = (market) => ({
  type: CHANGE_MARKET_NAME,
  payload: market,
});

export const fetchForexSuccess = (chart) => ({
  type: FETCH_FOREX_SUCCESS,
  payload: chart,
});

export const fetchChartError = (error) => ({
  type: FETCH_CHART_ERROR,
  payload: error,
});

export const fetchForex = (market) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/quote/${market}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    // FinancialModel API doesn't send error if it's about limit reach.
    // In that sense, they send an error message with 200 status. Weird.
    // Thats why I am catching that error here and dispatching it.
    if (data.length) {
      return data.length === 1
        ? [dispatch(setSearchMarketDetail(data))]
        : dispatch(fetchForexSuccess(data));
    } else {
      return dispatch(fetchChartError(data));
    }
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchChartData = (symbol, timeFrame) => async (dispatch) => {
  dispatch(cleanChartData());
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/historical-chart/${timeFrame}/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    return dispatch(setChartData(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchRating = (symbol) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/rating/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = await response.data;
    return dispatch(setRating(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

/*
This deserves to be mentioned.
Few months ago, when I was making the first version of this app, I used array.map to fetch charts.
Because of the async nation and using array map method, response was coming irregular and I had to sort them.
Now, with using for of loop, I dont need to write any extra code because loop waits for one to finish to move on.
Plus, its so clean!
Yet another homerun for me! 😁😎
*/
export const fetchHomeChart = (symbols) => async (dispatch) => {
  const symbolsArray = symbols.split(",").sort();
  let charts = [];
  try {
    // dispatch(setLoadingTrue());
    dispatch(cleanState());
    for (let symbol of symbolsArray) {
      const response = await axios.get(
        `https://financialmodelingprep.com/api/v3/historical-chart/1hour/${symbol}?apikey=${process.env.REACT_APP_CHART_KEY}`
      );
      const data = response.data;
      charts.push(data);
    }
    return charts.length
      ? dispatch(setHomeChartData(charts))
      : dispatch(fetchChartError(charts));
    // dispatch(setLoading());
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};

export const fetchSearch = (input) => async (dispatch) => {
  try {
    const response = await axios.get(
      `https://financialmodelingprep.com/api/v3/search?query=${input}&limit=10&apikey=${process.env.REACT_APP_CHART_KEY}`
    );
    const data = response.data;
    return dispatch(getSearchResults(data));
  } catch (error) {
    return dispatch(fetchChartError(error));
  }
};
