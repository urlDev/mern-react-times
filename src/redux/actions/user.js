import axios from 'axios';
import toaster from 'toasted-notes';
import { push } from 'connected-react-router';

import NotificationComponent from '../../components/notification-component/NotificationComponent';

const url = process.env.REACT_APP_API_URL;

export const REGISTER_USER = 'REGISTER_USER';
export const LOGOUT_USER = 'LOGOUT_USER';
export const LOGIN_USER = 'LOGIN_USER';
export const DELETE_USER = 'DELETE_USER';
export const UPDATE_USER = 'UPDATE_USER';
export const GET_USER = 'GET_USER';
export const SET_TOKEN = 'SET_TOKEN';
export const USER_FETCH_ERROR = 'USER_FETCH_ERROR';
export const USER_LOADING = 'USER_LOADING';
export const USER_LOADING_END = 'USER_LOADING_END';
export const USER_MODAL_OPEN = 'USER_MODAL_OPEN';
export const USER_MODAL_CLOSE = 'USER_MODAL_CLOSE';
export const DELETE_MODAL_OPEN = 'DELETE_MODAL_OPEN';
export const DELETE_MODAL_CLOSE = 'DELETE_MODAL_CLOSE';
export const UPLOAD_AVATAR = 'UPLOAD_AVATAR';

export const registerUser = (user) => ({
  type: REGISTER_USER,
  payload: user,
});

export const logOutUser = () => ({
  type: LOGOUT_USER,
});

export const loginUser = (user) => ({
  type: LOGIN_USER,
  payload: user,
});

export const deleteUser = () => ({
  type: DELETE_USER,
});

export const updateUser = (user) => ({
  type: UPDATE_USER,
  payload: user,
});

export const getUser = (user) => ({
  type: GET_USER,
  payload: user,
});

export const setToken = (token) => ({
  type: SET_TOKEN,
  payload: token,
});

export const userFetchError = (error) => ({
  type: USER_FETCH_ERROR,
  payload: error,
});

export const userLoading = () => ({
  type: USER_LOADING,
});

export const userLoadingEnd = () => ({
  type: USER_LOADING_END,
});

export const userModalOpen = () => ({
  type: USER_MODAL_OPEN,
});

export const userModalClose = () => ({
  type: USER_MODAL_CLOSE,
});

export const deleteModalOpen = () => ({
  type: DELETE_MODAL_OPEN,
});

export const deleteModalClose = () => ({
  type: DELETE_MODAL_CLOSE,
});

export const uploadAvatar = (images) => ({
  type: UPLOAD_AVATAR,
  payload: images,
});

export const fetchRegisterUser = (user) => async (dispatch) => {
  dispatch(userLoading());
  try {
    const response = await axios.post(`${url}/profile/register`, user);
    const data = await response.data;
    return [
      dispatch(registerUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
      dispatch(push('/')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Welcome! Enjoy your time`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      dispatch(userLoadingEnd()),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchLogoutUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  const path = window.location.pathname;

  try {
    await axios.post(`${url}/profile/logout`, null, config);

    return [
      path.includes('profile') && dispatch(push('/home')),
      dispatch(logOutUser()),
      toaster.notify(
        () => <NotificationComponent text={'Buh-Bye!'} success={true} />,
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return dispatch(userFetchError(error));
  }
};

export const fetchLoginUser = (user) => async (dispatch) => {
  dispatch(userLoading());
  try {
    const response = await axios.post(`${url}/profile/login`, user);
    const data = await response.data;
    return [
      dispatch(loginUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
      dispatch(push('/home')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Welcome! Enjoy your time`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      dispatch(userLoadingEnd()),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

// To get the user,
// but because I added an or statement in user and token initial states,
// app will check if there is user and token in localStorage
// so there will be no need for this.
// Adding just in case.
export const fetchUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  dispatch(userLoading());
  try {
    const response = await axios.get(`${url}/profile`, config);
    const data = await response.data;
    return [
      dispatch(registerUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
    ];
  } catch (error) {
    return [dispatch(userFetchError(error)), dispatch(userLoadingEnd())];
  }
};

export const fetchUpdateUser = (user) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    const response = await axios.patch(`${url}/profile`, user, config);
    const data = await response.data;
    return [
      dispatch(updateUser(data)),
      localStorage.setItem('user', JSON.stringify(data)),
      dispatch(push('/')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Updated successfully!`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchDeleteUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    await axios.delete(`${url}/profile`, config);
    return [
      localStorage.clear(),
      dispatch(push('/')),
      dispatch(deleteUser()),
      toaster.notify(
        () => <NotificationComponent success={true} text={`Buh-bye!`} />,
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            success={false}
            text={`Something went wrong!`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchUploadAvatar = (input) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  try {
    const response = await axios.post(`${url}/profile/avatar`, input, config);
    const data = await response.data;
    return [
      dispatch(uploadAvatar(data)),
      localStorage.setItem('user', JSON.stringify(data)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Avatar is changed! Wow, that looks amazing! ;)'}
            success={true}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};
