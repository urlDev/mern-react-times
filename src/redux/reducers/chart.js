import {
    FETCH_FOREX_SUCCESS,
    FETCH_CHART_ERROR,
    CHANGE_MARKET_TYPE,
    SET_MARKET_DETAIL,
    SET_CHART_DATA,
    SET_CHART_TIME_FRAME,
    SET_RATING,
    SET_HOME_CHART_DATA,
} from '../actions/chart';

import chartSort from '../utils/chartSort';

const initialState = {
    loading: true,
    forex: [],
    error: null,
    // Indexes are first to load.
    marketType: '%5EGSPC,%5ERUA,%5EDJI,%5ENDX,%5EN225,%5EFTSE',
    marketDetail: {},
    chartData: [],
    chartTimeFrame: '1hour',
    rating: [],
    homeChartData: [],
};

const chartReducer = (state = initialState, action) => {
    switch (action.type) {
        case FETCH_FOREX_SUCCESS:
            return {
                ...state,
                forex: action.payload,
                loading: false,
            };
        case FETCH_CHART_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case CHANGE_MARKET_TYPE:
            return {
                ...state,
                marketType: action.payload,
            };
        case SET_MARKET_DETAIL:
            return {
                ...state,
                marketDetail: action.payload.data,
            };
        case SET_CHART_DATA:
            return {
                ...state,
                chartData: action.payload,
                loading: false,
            };
        case SET_CHART_TIME_FRAME:
            return {
                ...state,
                chartTimeFrame: action.payload,
            };
        case SET_RATING:
            return {
                ...state,
                rating: action.payload,
            };
        case SET_HOME_CHART_DATA:
            return {
                ...state,
                homeChartData: [...state.homeChartData, action.payload],
                loading: false,
            };
        default:
            return state;
    }
};

export default chartReducer;