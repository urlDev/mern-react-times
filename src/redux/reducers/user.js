import {
    REGISTER_USER,
    SET_TOKEN,
    USER_FETCH_ERROR,
    USER_LOADING,
    USER_LOADING_END,
} from '../actions/user';

const initialState = {
    loadingUser: false,
    errorUser: null,
    user: JSON.parse(localStorage.getItem('user')) || {},
    token: JSON.parse(localStorage.getItem('token')) || {},
};

const userReducer = (state = initialState, action) => {
    switch (action.payload) {
        case REGISTER_USER:
            return {
                ...state,
                user: action.payload,
            };
        case SET_TOKEN:
            return {
                ...state,
                token: action.payload,
            };
        case USER_FETCH_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case USER_LOADING:
            return {
                ...state,
                loading: true,
            };
        case USER_LOADING_END:
            return {
                ...state,
                loading: false,
            };
        default:
            return state;
    }
};

export default userReducer;