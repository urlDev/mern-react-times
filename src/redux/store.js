import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));