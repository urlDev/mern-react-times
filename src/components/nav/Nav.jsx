import { useDispatch } from 'react-redux';

import { changeHeader } from '../../redux/actions/news';

import SearchSrc from '../../assets/searchIcon.svg';
import LogoSrc from '../../assets/logo.svg';

import { NavContainer, Logo, SearchIcon } from './Nav.styles';

const Nav = () => {
  const dispatch = useDispatch();

  return (
    <NavContainer>
      <SearchIcon src={SearchSrc} alt="search icon" />

      <Logo to="/home">
        <img
          style={{ height: '25px' }}
          src={LogoSrc}
          alt="react times logo"
          onClick={() => dispatch(changeHeader('Home'))}
        />
      </Logo>

      <button>Sign in</button>
    </NavContainer>
  );
};

export default Nav;
