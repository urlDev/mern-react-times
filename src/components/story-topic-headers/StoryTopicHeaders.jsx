import moment from "moment";
import { useDispatch, useSelector } from "react-redux";

import { changeHeader, clearError } from "../../redux/actions/news";

import ResponsiveMenu from "../responsive-menu/ResponsiveMenu";
import ResponsiveMenuModal from "../responsive-menu-modal/ResponsiveMenuModal";

import {
  StoryTopicHeadersContainer,
  TopicContainer,
  Date,
  StyledLink,
} from "./StoryTopicHeaders.styles";

export const topics = [
  "Home",
  "Arts",
  "Business",
  "Health",
  "Opinion",
  "Politics",
  "Science",
  "Sports",
  "Technology",
  "Travel",
];

const StoryTopics = () => {
  const { width, responsiveMenu } = useSelector((news) => news.news);
  const dispatch = useDispatch();

  return (
    <StoryTopicHeadersContainer>
      <Date>
        <span>{moment().format("dddd | ")}</span>
        {moment().format("MMM DD, YYYY")}
      </Date>
      <TopicContainer>
        {width < 1024 ? (
          <>
            <ResponsiveMenu />
            {responsiveMenu ? <ResponsiveMenuModal /> : null}
          </> //
        ) : (
          <>
            {topics.map((topic) => {
              return (
                <StyledLink key={topic} to={`/${topic.toLowerCase()}`}>
                  <span
                    onClick={() => {
                      dispatch(changeHeader(topic.toLowerCase()));
                      dispatch(clearError());
                    }}
                  >
                    {topic}
                  </span>
                </StyledLink>
              );
            })}
          </> //
        )}
      </TopicContainer>
    </StoryTopicHeadersContainer>
  );
};

export default StoryTopics;
