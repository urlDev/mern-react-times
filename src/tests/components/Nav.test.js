import React from "react";
import { shallow, mount } from "enzyme";
import { BrowserRouter as Router } from "react-router-dom";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";
import { user } from "tests/fixtures/user";

import Nav from "components/nav/Nav";
import Search from "components/search/Search";
import StoryTopicHeaders from "components/story-topic-headers/StoryTopicHeaders";
import { CLOSE_SEARCH_MODAL } from "redux/actions/chart";

// Test it with width smaller than 768 and bigger than 768
// Check if StoryTopicHeaders and Search has been rendered

describe("Testing Nav component", () => {
  let wrapper;
  let store;

  beforeEach(() => {
    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());

    jest
      .spyOn(ReactReduxHooks, "useDispatch")
      .mockImplementation(() => store.dispatch);

    Date.now = jest.fn(() => new Date("2021-01-16T00:00:00.000Z"));
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  describe("Testing the component without user components", () => {
    beforeEach(() => {
      store = mockStore({
        user: {},
        width: 900,
      });

      wrapper = shallow(<Nav store={store} />);
    });

    test("Should render the component without user components", () => {
      expect(wrapper).toHaveLength(1);
    });
    test("Should match snapshot, width > 900", () => {
      expect(wrapper).toMatchSnapshot();
    });

    test("Should render Search component", () => {
      // when width is more than 768, Search component renders
      expect(wrapper.containsMatchingElement(<Search />)).toEqual(true);
    });
    test("Should close search modal on button click", () => {
      wrapper.find("Link").childAt(0).simulate("click");

      const actions = store.getActions();
      expect(actions).toEqual([{ type: CLOSE_SEARCH_MODAL }]);
    });
  });

  describe("Should render the component with user components and state", () => {
    beforeEach(() => {
      store = mockStore({
        user,
        width: 700,
      });

      wrapper = mount(
        <Router>
          <Nav store={store} />
        </Router>
      );
    });
    test("Should render the component", () => {
      expect(wrapper).toHaveLength(1);
    });
    test("Should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
    test("Should get the state", () => {
      const state = store.getState();

      expect(state).toEqual({ user, width: 700 });
    });
    test("Should render StoryTopicHeader component, width < 768", () => {
      expect(wrapper.containsMatchingElement(<StoryTopicHeaders />)).toEqual(
        true
      );
    });
  });
});
