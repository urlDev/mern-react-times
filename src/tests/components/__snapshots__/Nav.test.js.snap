// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Nav component Should render the component with user components and state Should match snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Nav
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <styled.nav>
        <nav
          className="sc-iqHYGH dtgdtG"
        >
          <StoryTopicHeaders>
            <styled.div>
              <div
                className="sc-hBEYos bytodC"
              >
                <styled.p>
                  <p
                    className="sc-fFubgz jJqmWy"
                  >
                    <span>
                      Friday | 
                    </span>
                    Jan 15, 2021
                  </p>
                </styled.p>
                <styled.div>
                  <div
                    className="sc-fodVxV bDQCfy"
                  >
                    <ResponsiveMenu>
                      <styled.div
                        alt="hamburger menu"
                        onClick={[Function]}
                      >
                        <div
                          alt="hamburger menu"
                          className="sc-kstrdz cEwGfu"
                          onClick={[Function]}
                        >
                          <img
                            alt="hamburger icon"
                            src="hamburger.svg"
                          />
                        </div>
                      </styled.div>
                    </ResponsiveMenu>
                  </div>
                </styled.div>
              </div>
            </styled.div>
          </StoryTopicHeaders>
          <LogoComponent>
            <Styled(Link)
              to="/home"
            >
              <Link
                className="sc-jSgupP iwtquC"
                to="/home"
              >
                <LinkAnchor
                  className="sc-jSgupP iwtquC"
                  href="/home"
                  navigate={[Function]}
                >
                  <a
                    className="sc-jSgupP iwtquC"
                    href="/home"
                    onClick={[Function]}
                  >
                    <img
                      alt="react times logo"
                      onClick={[Function]}
                      style={
                        Object {
                          "height": "16px",
                        }
                      }
                    />
                  </a>
                </LinkAnchor>
              </Link>
            </Styled(Link)>
          </LogoComponent>
          <NavUserImage>
            <styled.div>
              <div
                className="sc-dQppl cYcFwf"
              >
                <styled.button
                  onClick={[Function]}
                >
                  <button
                    className="sc-bqyKva hGnPvf"
                    onClick={[Function]}
                  >
                    <UserImage
                      border="29px"
                      margin="10px"
                    >
                      <styled.picture
                        border="29px"
                        margin="10px"
                      >
                        <picture
                          className="sc-kEjbxe fwsIeO"
                        >
                          <source
                            srcSet="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAZAAAAGQCAYAAACAvzbMAAAACX"
                            type="image/png"
                          />
                          <img
                            alt="avatar"
                            src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAZAAAAGQCAYAAACAvzbMAAAACX"
                          />
                        </picture>
                      </styled.picture>
                    </UserImage>
                  </button>
                </styled.button>
              </div>
            </styled.div>
          </NavUserImage>
        </nav>
      </styled.nav>
    </Nav>
  </Router>
</BrowserRouter>
`;

exports[`Testing Nav component Testing the component without user components Should match snapshot, width > 900 1`] = `
<styled.nav>
  <Search />
  <LogoComponent />
  <Link
    style={
      Object {
        "textDecoration": "none",
      }
    }
    to="/profile/login"
  >
    <styled.button
      onClick={[Function]}
    >
      Login
    </styled.button>
  </Link>
</styled.nav>
`;
